### PR TITLE
Returning tokenGroups attribute as SID string instead of byte array

### LINKED ISSUE #61173: LDAP User Metadata: Return tokenGroups attribute as SID string, not byte array
When adding additional metadata fields to the Active Directory realm configuration, would it be possible for the **tokenGroups** attribute to be returned in SID string format so it could be used more easily in document queries for document level security?

Currently, if I add `metadata: ["tokenGroups"]` to `xpack.security.authc.realms.active_directory.ad_realm1` in elasticsearch.yml, querying `/_security/_authenticate` gives a response of string encoded bytes. When putting together the document query, I assume the `_user.metadata.tokenGroups` field looks similar and I don't believe there's a way to be able to compare this metadata to a field that uses SID format for representing permissions.

It looks like Elastic uses the **tokenGroups** user attribute to populate the "ldap_groups" and there is already [a method to convert the bytes to SID](https://github.com/elastic/elasticsearch/blob/627ef279fd29f8af63303bcaafd641aef0ffc586/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectorySIDUtil.java#L31) in the Elastic code base. I could see it being a modified version of the **ActiveDirectoryGroupsResolver** or a modification to the **LdapMetaDataResolver**, but I'm interested if this seems like a worthwhile addition to make it easier to compare user access in SID form to index SID fields.

Thanks!

### PR DESCRIPTION
**Pull request by BigPandaToo**
Returning tokenGroups attribute as SID string instead of byte array (AD metadata)

Resolves: #61173

### CODE REVIEW

**Review by jkakavas — COMMENTED** [review]
We should add a test for this that would fail without the changes before merging. Something in `ActiveDirectoryGroupsResolverTests` would make sense where we assert that the resolved groups contain the SID as a string

**Comment by BigPandaToo** [comment]
> We should add a test for this that would fail without the changes before merging. Something in `ActiveDirectoryGroupsResolverTests` would make sense where we assert that the resolved groups contain the SID as a string

Added the test

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java
Comment (jkakavas) [inline x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java:158]: This method looks very specific to the added test and its name is rather generic. I suggest we move the code to `testResolveTokenGroupsSID` , WDYT?

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java
Comment (jkakavas) [inline x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java:78]: `groupSIDs` -- > metadata or something similar that indicates that this the user metadata ( that contain tokenGroups )

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java
Comment (jkakavas) [inline x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java:68]: I think my original suggestion as to where we should put this test was misleading, sorry for that.  Probably `ActiveDirectorySessionFactoryTests` is a more proper place for this to live since it has to do with a specific attribute being properly added as metadata ( even though the attribute _is_ groups related )

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java:84]: How about we return a `String` here instead of a list with one item if tokenGroups contains only 1 value, similar to what we do for all other attributes ( see below ) ?

**Review by jkakavas — COMMENTED** [review]
Added some suggestions for changes

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java
Comment (BigPandaToo) [reply x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java]: Changed

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java
Comment (BigPandaToo) [reply x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java]: Done

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java
Comment (BigPandaToo) [reply x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolverTests.java]: Moved

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapMetadataResolver.java:85]: Done

File: x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectorySessionFactoryTests.java
Comment (jkakavas) [inline x-pack/qa/third-party/active-directory/src/test/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectorySessionFactoryTests.java:393]: Since we're not using a bind user in the settings we should probably use an ldap session that is created for the user that we search foe and resolve its attributes, so use `hulk` here instead

**Review by jkakavas — APPROVED** [review]
Thanks for the iteration @BigPandaToo. LGTM after the last few minor comments are addressed

**Pull request merged**
